incore_auth/app.py

```diff
@@ -40,6 +47,53 @@
     app.logger.setLevel(gunicorn_logger.level)
 
 
+def cache_key(request_info):
+    """Return username from request_info to be used as cache key"""
+    return request_info["username"]
+
+
+@cached(cache=TTLCache(maxsize=cache_size, ttl=cache_timeout), key=cache_key)
+def update_services(request_info):
+    """When a user does any action, it will check to update the groups in mongo, as well as make sure
+    the user has access to datawolf. The function is cached to make sure only every 30 minutes we do
+    the checks, since this can be expensive."""
+
+    # get information from request
+    username = request_info["username"]
+    if not username:
+        return username
+    groups = request_info['groups']
+
+    # call datawolf to add user
+    # TODO add datawolf call
+
+    # update database with user quota
+    mongo_client = config["mongo_client"]
+    if mongo_client:
+        mongo_user = mongo_client["spacedb"]["UserGroups"].find_one({"username": username})
+
+        if mongo_user is None:
+            # INSERT
+            mongo_client["spacedb"]["UserGroups"].insert_one({
+                "username": username,
+                "className": "edu.illinois.ncsa.incore.common.models.UserGroups",
+                "groups": groups
+            })
+            app.logger.debug(f"Inserted groups document for {username}")
```

Comment:
Could you also add logic to initiate a user space record if it doesnt' exist already

`mongo_space = mongo_client["spacedb"]["Space"].find_one({"metadata.name": username})`

```
{
    "className" : "edu.illinois.ncsa.incore.common.models.Space",
    "metadata" : {
        "className" : "edu.illinois.ncsa.incore.common.models.SpaceMetadata",
        "name" : "sanjaya"
    },
    "privileges" : {
        "className" : "edu.illinois.ncsa.incore.common.auth.Privileges",
        "userPrivileges" : {
            "sanjaya" : "ADMIN"
        }
    },
    "members" : [
    ]
}
```
You can repalce sanjaya with the current username

Reply:
done